### DISCUSSION TITLE
How to bootstrap a FutureProvider with a cached value?

### DISCUSSION THREAD

**Discussion by derolf** [rrousselGit/riverpod#2105] [Q&A]
I have a FutureProvider that calls an API to get some data. However, I also store cached values persistently to fill data on the next app start.

I am using the following simplified construct to bootstrap the FutureProvider with a cached value, but I dunno if that's the best way to do it:

```Dart
var String? cachedValue;

final dataProvider = FutureProvider(
  (ref) async {
    if (ref.state.isLoading && !ref.state.isRefreshing && !ref.state.isReloading) {
      if (cachedValue != null) {
        ref.state = AsyncData(cachedValue);
      }
    }
    final value = await ref.watch(apiProvider).getData();
    cachedValue = value;
    return value;
  },
);
```

EDIT (after Remi's comment):

In my app, `cachedValue` is of course not a global variable, but stored in a database.

**Comment by rrousselGit** [comment]
You shouldn't have a global variable like that.

`cachedValue` should be computed within your provider

**Reply by derolf** [reply]
> You shouldn't have a global variable like that.
> 
> `cachedValue` should be computed within your provider

It's only a global variable in the example for illustrative purposes. In my app, the cached data is stored in a database.

**Reply by rrousselGit** [reply]
As don't as the database read/write is done within your provider, then that's fine.

**Reply by derolf** [reply]
Do you have any other construct in mind that would be more suitable or am I doing it the canonical way?

**Reply by rrousselGit** [reply]
What you're doing is fine.
If you want something else, it'll have to wait until Riverpod has built-in offline support.
